### DISCUSSION TITLE
Compute average value of Elements within a certain height range of mesh

### DISCUSSION THREAD

**Discussion by hokonkim** [idaholab/moose#20946] [Q&A General] [answered]
Hello!

I have a question about computing average value of Elements within a certain height range of mesh.

For example, the mesh size I used for the simulation is 10 x 10 x 10 um^3.

I would like to compute the average value of Elements within the range of height from 0 to 1 um and export the value as output(*.CSV).

Is this possible?

If not, what would be the alternative?

Thank you for your help.

Best regards,
Hokon Kim

**Comment by GiudGiud** [comment]
Hello

So the ElementValueSampler is definitely what you will want to use
https://mooseframework.inl.gov/source/vectorpostprocessors/ElementValueSampler.html

You could create a new block for the region of interest and block-restrict it
For example with this: https://mooseframework.inl.gov/source/meshgenerators/ParsedSubdomainMeshGenerator.html

OR
you could just output it for the whole mesh and sort out the elements in the CSV file. There is even a sort_by parameter in the sampler, so you could sort the output by z then read only part of the file in your postprocessing workflow

Guillaume

**Reply by hokonkim** [reply]
Thanks for your reply.

I checked the first way you suggested. I got the point.

Specifically, what I want to compute is "one of Material Property values".

I don't think this Material Property can be dealt as Variables (or AuxVariables).

In ElementValueSampler, it looks like only Variable(AuxVariable) could be used.

Is there a way to copy Material Property into AuxVariable?

**Reply by GiudGiud** [reply]
there is
https://mooseframework.inl.gov/source/auxkernels/MaterialRealAux.html

**Reply by hokonkim** [reply]
I really appreciate your help!

I would try this way, and let's see!

**Reply by hokonkim** [reply]
I have questions.

First of all, when I made a subdomain through ParsedSubdomainMeshGenerator, subdomain requires Active Kernel.

I thought that subdomain restricted the interested domain including the information I assigned at the original blocks, but it didn't look like that.

Is there a way making a subdomain copying the information from the original block?
(with no requirement for Active Kernel, just used for computing the average value of Elements)



Also, when I tried ElementValueSampler, many csv files for each time step were exported.

Ultimately what I want to print out in csv file is "the single value representing the average value of Elements within a certain range of height" like below.

Voltage | Current | ........ | Avg. of K from 0 to 1 um | Avg. of K from 1 to 2 um | ... |
    1.02   |   3.1      | ........ |                     2034            |                   1200               | ... |

If I use ElementValueSampler, is there a way not to generate other csv files and to print out the information into original csv file?

**Reply by GiudGiud** [reply]
Hello

So you can turn off the requirement for an active kernel with this option
```
[Problem]
  kernel_coverage_check = false
[]
```

But it's better to just not block restrict the kernels and variables and let them execute everywhere in your case I think. That way the subdomain modifications wont affect the equations. 

You can use this option `contains_complete_history` to merge all the csv files for each time step with the `ElementValueSampler`. 
You cannot merge the postprocessors and the vectorpostprocessor output that easily. So this table you describe would only work if you use the subdomain methods and not the ElementValueSampler imo

**Reply by hokonkim** [reply]
Then, what about this combinations?

Several subdomains (by using ParsedSubdomainMeshGenerator)
- I will assign the needed parameters like the target Elemental value(Materials property)
- Kernels will work on each subdomain similar to original blocks.

ElementAverageMaterialProperty
- This postprocessor will work on each subdomain and give me the single value (average) and export the value in CSV.

I just attached the code and mesh for simple test. But after assigning parameters into subdomain, I cannot get the result.
[TEST.zip](https://github.com/idaholab/moose/files/8637468/TEST.zip)

**Reply by GiudGiud** [reply]
What app do I need to run this? Looks like you have some custom objects. 

It seems you have added sub_0 to all the block restrictions in the system. 
Can you add show_info = true to the ParsedSubdomainMeshGenerator ? To see if it worked
Also what are you getting for the result currently?

**Reply by hokonkim** [reply]
Yes, I forgot that in my apps, there were several custom kernels and materials.

"show_info = true" in ParsedSubdomainMeshGenerator gave me some information about the subdomain.

I just knew that the subdomain includes unnecessary blocks' information.

Currently, I use "combinatorial_geometry = 'z > 0 & z < 0.0002 & x > 1.89861e-5 & x < 0.0002 & y > 1.89939e-5 & y < 0.0002'"

However, what I want to do is a certain height range of one block ('PT_TPB_TET4'), but above rectangle includes a part of 4 block's information.

Is there a way to pick a certain block as subdomain? not the rectangle..
(e.g., z in 'PT_TPB_TET4' > 0 & z in 'PT_TPB_TET4' < 0.0002)

Lastly, if I add subdomain into the current input file, then it can be run, but it cannot converge.
(I think the subdomain has the part of information, specifically the subdomain is disconnected so it cannot converge)

**Reply by GiudGiud** [reply]
Try adding the block = '... sub_0' in the GlobalParams. You're probably missing a block-restriction in one object. 

I dont think we can do what you want with the mesh generators right now.
Try this:
- place z in an auxvariable using a ParsedFunction and a FunctionAux
- use the https://mooseframework.inl.gov/source/userobject/CoupledVarThresholdElementSubdomainModifier.html with z as the variable and the threshold and block-restrict it to PT_TPB_TET4. If you only want the final result, you can execute this on FINAL and not have to worry about block restriction of the kernels
- use the ElementValueSampler with block-restriction on the new block

**Reply by hokonkim** [reply]
[Functions]
      [./testFunction]
        type = ParsedFunction
        value = 'z > 0 & z < 0.0002'
      [../]
    []

    [UserObjects]
      [./just_test]
        type = CoupledVarThresholdElementSubdomainModifier
        coupled_var = 'test_z'
        block = 'PT_TPB_TET4'
        threshold = 0
        subdomain_id = 100
      [../]
    []

    [AuxVariables]
      [./test_z]
        order = CONSTANT
        family = MONOMIAL
        block = 'PT_TPB_TET4'
      [../]
    []

    [AuxKernels]
      [./a]
        type = FunctionAux
        variable = test_z
        function = testFunction
      [../]
    []


I don't get it how z can be used for specifying the interested height.

Also, I don't know how to adjust threshold and sub_domain_id in CoupledVarThresholdElementSubdomainModifier.
I thought "sub_domain_id" is used to specify the new block I'm interested.

"Requested subdomain 100 does not exist" is the error message I got for now....

Does this new subdomain affect the original mesh ('PT_TPB_TET4')?

**Comment by GiudGiud** [comment, marked as answer]
try this
```
[Functions]
  [.z_function]
    type = ParsedFunction
    value = 'z'
  [../]
[]

[UserObjects]
  [./just_test]
    type = CoupledVarThresholdElementSubdomainModifier
    coupled_var = 'z_var'
    block = 'PT_TPB_TET4'
    threshold = 0.0002
    subdomain_id = 100
    execute_on = FINAL
  [../]
[]

[AuxVariables]
  [z_var]
    order = CONSTANT
    family = MONOMIAL
    block = 'PT_TPB_TET4'
  [../]
[]

[AuxKernels]
  [./a]
    type = FunctionAux
    variable = z_var
    function = z_function
  [../]
[]
```

**Reply by hokonkim** [reply]
[Postprocessors]
      [./TEST]
        type = ElementAverageMaterialProperty
        mat_prop = 'tpbActivity_S0'
        block = '100'
      [../]
    []

The codes you gave me worked.

I added the PP like above, but the error message "the following blocks (ids) do no exist on the mesh: 100" popped up.


I tried this, but it showed the same error as well.

    [VectorPostprocessors]
      [./test]
        type = ElementValueSampler
        block = '100'
        sort_by = z
        variable = z_var
      [../]
    []


[testcode.zip](https://github.com/idaholab/moose/files/8643420/testcode.zip)

**Reply by GiudGiud** [reply]
when you look at the mesh does the subdomain appear?

Maybe I did not set the threshold right?

**Reply by GiudGiud** [reply]
also the UserObject is executed on FINAL, so before that if you run anything on subdomain 100, it does not exist yet

**Reply by hokonkim** [reply]
First, in the block, there were only 4 blocks. subdomain didn't appear.

I ran the UserObject without applying execute_on.

**Reply by GiudGiud** [reply]
then you need to check the parameters.

btw for the previous approach, there is a parameter to exclude some subdomains 
https://mooseframework.inl.gov/source/meshgenerators/ParsedSubdomainMeshGenerator.html
I didnt know about it

**Reply by hokonkim** [reply]
<img width="1040" alt="Screen Shot 2022-05-06 at 5 51 57 PM" src="https://user-images.githubusercontent.com/46009457/167221255-b9a78c33-a461-4803-bed8-4308ff96c826.png">

Currently, I'm still stuck.

Based on what you suggested, I will try to implement the function that I wanted.

For exclude some subdomain parameter, there is no way to specify the other blocks inside of subdomain as far as I know....

**Reply by GiudGiud** [reply]
You specify the domains you exclude, not the domains you include. 

Yup paraview confirms the user object did not add the block.
Can you execute the user object and the FunctionAux on INITIAL and TIMESTEP_BEGIN and see if it works then?
The z_var needs to be populated and the UO needs to be exceuted before other code referring to them is ran
